src/server/qgis_mapserver.cpp

```diff
@@ -357,6 +348,23 @@ int main( int argc, char *argv[] )
             }
           }
 
+          // Build URL from env ...
+          QString url { qgetenv( "REQUEST_URI" ) };
+          // ... or from server ip/port and request path
+          if ( url.isEmpty() )
+          {
+            const QString path { firstLinePieces.at( 1 )};
+            // Take Host header if defined
+            if ( headers.contains( QStringLiteral( "Host" ) ) )
+            {
+              url = QStringLiteral( "http://%1%2" ).arg( headers.value( QStringLiteral( "Host" ) ) ).arg( path );
+            }
+            else
+            {
+              url = QStringLiteral( "http://%2:%3%4" ).arg( ipAddress ).arg( port ).arg( path );
```

Comment:
shouldn't that be "http://%1:%2%3"  ?

Reply:
oh, yes of course, thanks

Reply:
would be good if there was a tool to detect this. gcc can detect bad arguments passed to good old printf() :-)

Reply:
Yeah, I think all we get from Qt is a runtime warning.

Reply:
Not even, it will just work as expected. After "fixing" it you'll have

`QString( "http://%1:%2%3" ).arg( ipAddress )` -> returns `QString( "http://127.0.0.1:%2%3" )`
`QString( "http://127.0.0.1:%2%3" ).arg( port )` -> ...

see https://doc.qt.io/qt-5/qstring.html#arg

> Returns a copy of this string with the lowest numbered place marker replaced by string a, i.e., %1, %2, ..., %99.

/smart-arse mode off